### ISSUE TITLE
Suggested code does not compile with or_fun_call

### ISSUE THREAD

**Issue by jDomantas** [rust-lang/rust-clippy#2186]
In some cases suggested change regarding `or_fun_call` lint does not compile. Here's an example:

```rust
struct Pair {
    a: u32,
    b: u32,
}

fn foo(x: u32) -> u32 { x }

fn bar(pair: &mut Pair) {
    let borrow = &mut pair.a;
    let b = None.unwrap_or(foo(pair.b));
}
```

`pair` is partially borrowed. Calling `id(pair.b)` is valid, because compiler can see that only field `a` is borrowed, so field `b` can be freely used. Clippy suggests to use `unwrap_or_else`, and move function call into a closure. However, the closure will borrow `pair` fully which is not possible, and the code will not compile.

**Comment by lzutao** [comment]
I believed NLL solved this.

**Comment by jDomantas** [comment]
This specific example is only solved just because `borrow` is not used later, you only need to add a use after `unwrap_or` to make the suggestion incorrect again:

```rust
struct Pair {
    a: u32,
    b: u32,
}

fn foo(x: u32) -> u32 { x }

fn bar(pair: &mut Pair) {
    let borrow = &mut pair.a;
    let b = None.unwrap_or(foo(pair.b));
    let x = *borrow;
}
```

**Comment by lzutao** [comment]
I think with this RFC (https://rust-lang.github.io/rfcs/2229-capture-disjoint-fields.html) the code should be accepted and compiled.
